cou/apps/app.py

```diff
@@ -182,7 +305,251 @@ def _get_os_origin(self) -> str:
         """
         for origin in ("openstack-origin", "source"):
             if self.config.get(origin):
+                self.origin_setting = origin
                 return self.config[origin].get("value", "")
 
         logger.warning("Failed to get origin for %s, no origin config found", self.name)
         return ""
+
+    async def _check_upgrade(self, target: OpenStackRelease) -> None:
+        """Check if an application has upgraded its workload version.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        """
+        status = await async_get_status()
+        app_status = status.applications.get(self.name)
+        units_not_upgraded = []
+        for unit in app_status.units.keys():
+            workload_version = app_status.units[unit].workload_version
+            compatible_os_versions = OpenStackCodenameLookup.lookup(self.charm, workload_version)
+            if target not in compatible_os_versions:
+                units_not_upgraded.append(unit)
+        if units_not_upgraded:
+            logger.error(
+                "App: '%s' has units: '%s' didn't upgrade to %s",
+                self.name,
+                ", ".join(units_not_upgraded),
+                str(target),
+            )
+
+    def pre_upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Pre Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add pre upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        return [self._get_refresh_charm_plan(target)]
+
+    def upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation
+        :return: Plan that will add upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        if self.current_os_release >= target:
+            logger.warning(
```

Comment:
it may be more useful to say which version the app is actually running. Also, `info` may be more appropriate in case we end up here after resuming an upgrade.

Reply:
Changed to info